pallets/pools/src/solution.rs

```diff
@@ -0,0 +1,649 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use super::*;
+
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug, TypeInfo)]
+pub enum PoolState {
+	Healthy,
+	Unhealthy(Vec<UnhealthyState>),
+}
+
+impl PoolState {
+	/// Updates a PoolState to update.
+	///
+	/// NOTE:
+	/// * This will switch a PoolState::Healthy -> PoolState::Unhealthy(_) and vice versa
+	/// * If an already unhealthy state is updated, the new Vec<UnhealthyState> inside the
+	///   enum will be **overwritten** with the newly passed unhealthy states.
+	///   -> Use `add_unhealthy` or `rm_unhealthy` if the other states should be kept.
+	pub fn update(&mut self, update: PoolState) -> &mut Self {
+		*self = update;
+		self
+	}
+
+	pub fn add_unhealthy(&mut self, add: UnhealthyState) -> &mut Self {
+		match self {
+			PoolState::Healthy => {
+				let mut states = Vec::new();
+				states.push(add);
+				*self = PoolState::Unhealthy(states);
+				self
+			}
+			PoolState::Unhealthy(states) => {
+				if !states.contains(&add) {
+					states.push(add);
+				}
+				self
+			}
+		}
+	}
+
+	pub fn rm_unhealthy(&mut self, rm: UnhealthyState) -> &mut Self {
+		match self {
+			PoolState::Healthy => {
+				let mut states = Vec::new();
+				states.push(rm);
+				*self = PoolState::Unhealthy(states);
+				self
+			}
+			PoolState::Unhealthy(states) => {
+				states.retain(|val| val != &rm);
+
+				if states.len() == 0 {
+					*self = PoolState::Healthy;
+				}
+				self
+			}
+		}
+	}
+}
+
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug, TypeInfo)]
+pub enum UnhealthyState {
+	MaxReserveViolated,
+	MinRiskBufferViolated,
+}
+
+/// The solutions struct for epoch solution
+#[derive(Encode, Decode, Clone, Eq, RuntimeDebug, TypeInfo)]
+pub enum EpochSolution<Balance> {
+	Healthy(HealthySolution<Balance>),
+	Unhealthy(UnhealthySolution<Balance>),
+}
+
+impl<Balance> EpochSolution<Balance>
+where
+	Balance: Copy,
+{
+	pub fn healthy(&self) -> bool {
+		match self {
+			EpochSolution::Healthy(_) => true,
+			EpochSolution::Unhealthy(_) => false,
+		}
+	}
+
+	pub fn solution(&self) -> &[TrancheSolution] {
+		match self {
+			EpochSolution::Healthy(solution) => solution.solution.as_slice(),
+			EpochSolution::Unhealthy(solution) => solution.solution.as_slice(),
+		}
+	}
+}
+
+impl<Balance> PartialEq for EpochSolution<Balance>
+where
+	Balance: PartialEq,
+{
+	fn eq(&self, other: &Self) -> bool {
+		match self {
+			EpochSolution::Healthy(s_1) => match other {
+				EpochSolution::Healthy(s_2) => s_1 == s_2,
+				EpochSolution::Unhealthy(_) => false,
+			},
+			EpochSolution::Unhealthy(s_1) => match other {
+				EpochSolution::Healthy(_) => false,
+				EpochSolution::Unhealthy(s_2) => s_1 == s_2,
+			},
+		}
+	}
+}
+
+impl<Balance> PartialOrd for EpochSolution<Balance>
+where
+	Balance: PartialOrd,
+{
+	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		match self {
```

Comment:
Since the `Healthy` variant comes first we need to be explicit here, but we could list `Unhealthy` first and just use derive: https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html#derivable

If we do want to keep being explicit, it would be nice to collapse this to a single `match (self, other)`

Reply:
I want to keep being explicit about the `PartialOrd` just for the sake of readability and understandability. But I will collapse into a single match.